File: 프로그래머스/lv2/권민서/올바른괄호.java
import java.util.*;

class Solution {
    public boolean solution(String str){
        Stack<Character> stack = new Stack<>();
        boolean answer = false;
        for(char x : str.toCharArray()) {
            if(x == '(') stack.push(x);
            else {
                if(stack.isEmpty()) return answer;
                stack.pop();
            }
        }
        if(stack.isEmpty()) answer = true;
        return answer;
    }

    public static void main(String[] args) {
        Solution T = new Solution();
        Scanner kb = new Scanner(System.in);
        String str = kb.next();
        T.solution(str);
    }

}

File: 프로그래머스/lv2/올바른 괄호.cpp
#include<string>
#include <iostream>
#include <stack>
using namespace std;

bool solution(string s)
{
    stack<int> st;
    
    for( char c : s ) {
        if(c=='(') {
            st.push(1);
        }
        if(c==')') {
            if(st.empty()) {
                return false;  //단순히 갯수만 맞는게 아니라 )가 먼저 오는 경우가 있으면 안됨
            }
            else {
                 st.pop();
            }
        
        }
    }
    if(st.empty()) {
        return true;
    }
    else {
        return false;
    }

}
